### PR TITLE
Immersive telepresence

### PR DESCRIPTION
**Pull request by ZuzkaU**
Changes in TeleportVR

1. Server

There is a new abstract class ServerData, which hides the implementation details of the specific server. Other classes can access to its functions without being bound to Animus / WebRTC / mock.
The server implementation can be changed in the scene by enabling the correct GameObject.
Aiortc server implementation is added from the demo project.

2. Scene management

For scene switching, only a single class is used: SceneManager. After loading the Main scene, it automatically sets all object from it as "Dont destroy on load", and loads the training scene. It switches between training scene and the real world scene.
The Main scene in Unity should only contain objects that will persist between the scenes.
GameObjects should be related to only one component, which can be easily enabled / disabled through enabling / disabling the GameObject. For example, the eye planes / spheres to project the camera images to are created by the server component depending on which camera is used, and not present the whole time.

3. Widgets

Widgets are created from a function (in class WidgetFactory), and not from a template. The class BasicWidgets contains the main widgets shown on the top bar of HUD.

4. Input System

The class InputSystem allows to assign function delegates to named buttons, so that the other classes can abstract away the particular input device which is used. The implemented classes ControllerInputSystem and KeyboardInputSystem take care of the function invoke.

5. Folder structure

Changed folder structure to match standard structure in Unity projects. Moved the majority of the code to Assets/Scripts. Gathered code of various control devices to the folder InputDevices. Removed some unused files.

### CODE REVIEW

**Review by missxa — CHANGES_REQUESTED** [review]
- [ ] add inline comments with explanations to all newly created classes (at least)
- [x] use member class variables in functions called very often (eg from `Update()`) instead of initialising local variables, since it uses up a lot of memory, especially bad on Android
- [ ] address comments re refactoring & data connectors @ZuzkaU 
- [ ] address comments re new webrtc communication @bekleyis95 
- [x] remove unnecessary logs
- [ ] pass user tests by @missxa 
- [x] clean up `.meta~*` files
- [x] move all scripts related to Robody control to `RobodyControl` folder
- [x] move tutorial folder into the training folder
- [x] rename `Widgets` folder to `WidgetsBaseImplementation` or similar, since the name is misleading. For example, clock is a widget
- [x] update README

**Comment by ZuzkaU** [comment]
- [ ] fix centering the VR Rig on load (it gets centered but is 90 degrees rotated)
- [ ] use B button for recentering again
- [ ] make HUD transparent
- [ ] use the plugin for buttons
- [ ] remove rays from the controllers (only keep them for Settings)
- [ ] icons on the top bar: make clear which state they are in (mic on / off / grayed out if not connected / not in HUD?), add icon for movements
- [ ] don't display the spheres when the data from the camera haven't arrived yet
- [ ] use milky way skybox by default (we should have plugin for that)
- [ ] we need a server name for aiortc instead of a hardcoded IP address

File: Assets/Scripts/Tutorial/RightHand.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+namespace Tutorial
+{
+    /// <summary>
+    /// Close the right hand to grip.
+    /// </summary>
+    public class RightHand : Step
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Tutorial/RightHand.cs]: fac850728ec916a1fad4c0acf21

File: Assets/Scripts/Tutorial/MoveHead.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+using InputDevices;
+using UnityEngine;
+using Widgets;
+
+namespace Tutorial
+{
+    /// <summary>
+    /// Try to move head and nod.
+    /// </summary>
+    public class MoveHead : Step
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Tutorial/MoveHead.cs]: fac850728ec916a1fad4c0acf21

File: Assets/Scripts/Tutorial/LeftHand.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+namespace Tutorial
+{
+    /// <summary>
+    /// Close the left hand to grip.
+    /// </summary>
+    public class LeftHand : Step
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Tutorial/LeftHand.cs]: fac850728ec916a1fad4c0acf21

File: Assets/Scripts/DisplayExtensions/ClockWidget.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        public TextWidget timeWidget;
+        public void Start()
+        {
+            timeWidget = WidgetFactory.Instance.CreateTextWidget("what's the time?", WidgetPosition.TopRight, "clockWidget");
+        }
+
+        private int oldSeconds = 0;
+        public void Update()
+        {
+            string time = $"{DateTime.Now.Hour:D2}:{DateTime.Now.Minute:D2}:{DateTime.Now.Second:D2}";
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/DisplayExtensions/ClockWidget.cs:17]: https://docs.unity3d.com/Manual/performance-garbage-collection-best-practices.html

File: Assets/Scripts/Tutorial/LeftArm.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+using InputDevices.Controllers;
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Tutorial/LeftArm.cs]: fac850728ec916a1fad4c0acf216b33a321cf9a3

File: Assets/Scripts/Tutorial/Done.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+namespace Tutorial
+{
+    public class Done : Step
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Tutorial/Done.cs]: fac850728ec916a1fad4c0acf216b33a321cf9a3

File: Assets/Scripts/ServerConnection/Aiortc/aiortcConnector.cs [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    void Start()
+    {
+        leftRenderer = imtpEncoder.leftEye.GetNamedChild("LeftSphereMaterial").GetComponent<Renderer>();
+        leftFaceDetectionRenderer = imtpEncoder.leftEye.GetNamedChild("Plane").GetComponent<Renderer>();
+        leftFaceDetectionRenderer.material.mainTexture = new Texture2D(1080, 1080);
+        rightRenderer = imtpEncoder.rightEye.GetComponentInChildren<Renderer>();
+        
+        onDataChannel = channel =>
+        {
+            Debug.Log($"onDataChannel");
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/ServerConnection/Aiortc/aiortcConnector.cs]: dc7d98d55feef0fb79a3a39fb28b2f6e525c7406

File: Assets/Scripts/InputDevices/JointTransfer/DrawDebugObj.cs [hunk 8]
```diff
@@ -22,5 +22,21 @@ void Start()
                 cube.transform.localScale = new Vector3(scale, scale, scale);
             }
         }
+
+
+        private IEnumerable<GameObject> getChildren(Transform parent, System.Predicate<GameObject> predicate)
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/InputDevices/JointTransfer/DrawDebugObj.cs:27]: It was in this class originally, I moved the folder JointTransfer from Assets/InputDevices to Asssets/Scripts/InputDevices and did only little changes here.

https://github.com/Roboy/TeleportVR/commit/4380511f6bdcaf459972727d464714416f93f965#diff-b880860836c504b5f3ad0afd241f62acd076cfc6824f0c3637fd713d73cfd427

File: Assets/Scripts/InputDevices/Controllers/ControllerInputSystem.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Reflection;
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace InputDevices.Controllers
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/InputDevices/Controllers/ControllerInputSystem.cs]: c0b48d39e9c4f2a2150ce2cbc4a5d140254c47b4

File: Assets/Scripts/InputDevices/Scripts/HandManager.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
+using System.Collections;
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/InputDevices/Scripts/HandManager.cs]: Only the location. https://github.com/Roboy/TeleportVR/blob/553654a0e02cf72600ca52ed509bf30cd24f31aa/Assets/InputDevices/Scripts/HandManager.cs

File: Assets/Scripts/InputDevices/Scripts/XROffset.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
+using System.Collections;
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/InputDevices/Scripts/XROffset.cs]: 19ac860d32fccec00

File: Assets/Scripts/ServerConnection/Aiortc/ImtpEncoder.cs [hunk 12]
```diff
@@ -0,0 +1,135 @@
+using System;
+using ServerConnection;
+using UnityEngine;
+using UnityEngine.UI;
+
+public class ImtpEncoder : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/ServerConnection/Aiortc/ImtpEncoder.cs:11]: 2ac8647

File: Assets/Scripts/Widgets/BasicWidgets.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
+using ServerConnection;
+
+namespace Widgets
+{
+    /// <summary>
+    /// This class creates the most essential widgets at startup and manages their state.
+    /// </summary>
+    public class BasicWidgets : Singleton<BasicWidgets>
```
[the remark this answers is not in the document]
Comment (ZuzkaU) [reply Assets/Scripts/Widgets/BasicWidgets.cs:8]: This is *not* a widget manager. This is an implementation of a few widgets which are so essential that they should be present every time the Widgets component is in use. Also, if you name every other component "manager", it doesn't increase clarity of what it actually manages.

File: Assets/Scripts/DisplayExtensions/ClockWidget.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (missxa) [reply Assets/Scripts/DisplayExtensions/ClockWidget.cs]: @ZuzkaU  yes, this supports my comment above. or did you have any objections?

**Review by missxa — APPROVED** [review]

**Pull request merged**